conn_pool.go

```diff
@@ -105,7 +112,24 @@ func (p *ConnPool) Acquire() (c *Conn, err error) {
 		if p.logLevel >= LogLevelWarn {
 			p.logger.Warn("All connections in pool are busy - waiting...")
 		}
+
+		deadlineReached := false
+
+		if p.connAllocTimeout > 0 {
+			timer := time.NewTimer(p.connAllocTimeout)
+			defer timer.Stop()
+
+			select {
+			case <-timer.C:
+				deadlineReached = true
+				p.cond.Signal()
+			}
```

Comment:
When a connection is returned back to a pool, Signal() method is called, and every waiter gets unlocked. The waiters check if they could acquire a connection, and if not, the get blocked again.

In this code we manually call Signal() to unblock all the waiters to check whose timers got expired. Ideally only this waiter exits with a timeout and all the other go back to sleep.